Accessibility

blavalla opened the pull request:
Added the basic start for accessibility inspection, and a wrapper class to hold all of the methods for various accessibility properties I plan to add to the inspector.  For now these will appear as a second CSS Selector alongside other view properties.  The only one currently implemented is "ignored", which signifies whether or not a node will be ignored by accessibility services.

I plan to follow the properties of the AXNode type in the Chrome Debugging Protocol (http://chromedevtools.github.io/debugger-protocol-viewer/tot/Accessibility/#type-AXNode).  This will make it easier to transition to the Accessibility Panel if Stetho is upgraded past the 1.1 version of the protocol.

I'm not sure if this is a problem, but by adding the "getStyleFromBoolean" and getStyleFromCharString" methods into ViewDescriptor.java, I have exposed a lot more of the methods/fields of the standard View class.  If that's not desired, I can change the approach here.

Screenshot:
![screen shot 2016-05-31 at 2 33 01 pm](https://cloud.githubusercontent.com/assets/1518064/15691315/cad690b4-273c-11e6-967d-e24128a8072d.png)

blavalla commented:
Alright, this should be rebased with master so that the changes from #399 included.

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,53 @@
+    mView = view;
+    mParent = view.getParent();
+  }
+  @ViewDebug.ExportedProperty(category = "accessibility")
```
nit: line of whitespace between methods

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,53 @@
+public class AccessibilityNodeInfoWrapper {
+  private AccessibilityNodeInfoCompat mNodeInfo;
+  private View mView;
+  private ViewParent mParent;
```
can these be made `final` ? if so, please do so

blavalla commented:
- Added some short-circuits to AccessibilityNodeInfoWrapper.
- Removed a bunch of unnecessary reflection from ViewDescriptor in favor of simply calling the methods from AccessibilityNodeInfoWrapper.

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // AccessibilityNodeInfo has text set (commonly set from TextViews and Switches).
+    if (mNodeInfo.getText() != null && mNodeInfo.getText().length() > 0) {
```
does `getText()` allocate? if it does, or if we don't know, then we should call it once

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // AccessibilityNodeInfo has a content description
+    if (mNodeInfo.getContentDescription() != null &&
```
does `getContentDescription()` allocate? if it does, or if we don't know, then we should only call it once

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,65 @@
+import android.view.View;
+import android.view.ViewParent;
+
+public class AccessibilityNodeInfoWrapper {
```
Can we make this into a class with static utility methods? I'm not sure we need an allocation for the sake of a `getIgnored` method with 3 parameters (or even for when/if we add more methods)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,65 @@
+package com.facebook.stetho.inspector.elements.android;
```
new files need copyright added a top. (copy+paste from any other .java file.)

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/CSS.java:
```diff
@@ -99,8 +99,13 @@ public JsonRpcResult getMatchedStylesForNode(JsonRpcPeer peer, JSONObject params
+    matchedCSSRules.add(match);
+    matchedCSSRules.add(accessibilityMatch);
+
+    result.matchedCSSRules = ListUtil.copyToImmutableList(matchedCSSRules);
```
in this case, since you're always populating the `List` with 2 elements, just use `ListUtil.newImmutableList` directly. You'll need to add a 2 parameter version that instantiates `TwoItemImmutableList`.

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/AbstractChainedDescriptor.java:
```diff
@@ -155,4 +155,14 @@ public final void getStyles(Object element, StyleAccumulator accumulator) {
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public final void getAccessibilityNodeInfo(Object element, StyleAccumulator accumulator) {
```
maybe call this `getAccessibilityStyles` so it matches up with `getStyles`, and because we're still using `StyleAccumulator`. (and then follow up with the cascading renames elsewhere of course)

rickbrew commented:
Okay I added another round of notes. Looking pretty good, getting pretty close :)

blavalla commented:
Latest changes should all be addressed.  Thanks!

rickbrew commented on stetho/src/main/java/com/facebook/stetho/inspector/elements/android/AccessibilityNodeInfoWrapper.java:
```diff
@@ -0,0 +1,69 @@
+
+public final class AccessibilityNodeInfoWrapper {
+
+  public AccessibilityNodeInfoWrapper() {
```
prefer a private constructor, or none at all

private constructor helps to mimic "static classes" in C#/VB. not having an explicitly defined constructor is equivalent to having this public parameterless constructor, but is a bit cleaner for us here

rickbrew commented:
Awesome ... looks good to me 👍  Let me know when yo'ud like me to merge
